k4FWCore/components/IOSvc.cpp

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2014-2024 Key4hep-Project.
+ *
+ * This file is part of Key4hep.
+ * See https://key4hep.github.io/key4hep-doc/ for further info.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "IOSvc.h"
+
+#include "edm4hep/MCParticleCollection.h"
+#include "podio/Frame.h"
+#include "podio/FrameCategories.h"
+
+#include "k4FWCore/KeepDropSwitch.h"
+
+#include "GaudiKernel/AnyDataWrapper.h"
+#include "GaudiKernel/IEventProcessor.h"
+
+#include <mutex>
+#include <thread>
+#include <tuple>
+
+StatusCode IOSvc::initialize() {
+  if (!m_readingFileNames.empty()) {
+    m_reader = std::make_unique<podio::ROOTReader>();
+    try {
+      m_reader->openFiles(m_readingFileNames);
+    } catch (std::runtime_error& e) {
+      error() << "Error when opening files: " << e.what() << endmsg;
+      throw e;
+    }
+    m_entries = m_reader->getEntries(podio::Category::Event);
+  }
+
+  m_switch = KeepDropSwitch(m_outputCommands);
+
+  m_incidentSvc = service("IncidentSvc");
+  m_incidentSvc->addListener(this, IncidentType::EndEvent);
+
+  m_dataSvc = service("EventDataSvc");
+
+  m_hiveWhiteBoard = service("EventDataSvc");
+
+  return Service::initialize();
```

Comment:
I added a couple of checks; for m_hiveWhiteBoard I can't do a simple check because it's not always present; later on there is a check before using it. I added `Service::initialize()`.